pep-0646.rst

```diff
@@ -155,432 +95,344 @@ Type variable tuples are created with:
+within ``Tuple``:
 
 ::
 
-    def f(*args: *Ts) -> Union[*Ts, float]: ...
-
-    f(0, 'spam')  # Inferred type is Union[int, str, float]
+    def foo(*args: *Ts): ...  # NOT valid
 
-Concatenating Multiple Type Variable Tuples
--------------------------------------------
-
-We can also concatenate *multiple* type variable tuples, but only in cases
-where the types bound to each type variable tuple can be inferred
-unambiguously. Note that this is not always the case:
+Following `Type Variable Tuples Must Have Finite Length When Bound`, note
+that the following should *not* type-check as valid (even though it is, of
+course, valid at runtime):
 
 ::
 
-    # Type checker should raise an error on definition of func;
-    # how would we know which types are bound to Ts1, and which
-    # are bound to Ts2?
-    def func(ham: Tuple[*Ts1, *Ts2]): ...
-
-    # Ts1 = Tuple[int, str], Ts2 = Tuple[bool]?
-    # Or Ts1 = Tuple[int], Ts2 = Tuple[str, bool]?
-    ham: Tuple[int, str, bool]
-    func(ham)
+    def foo(*args: Tuple[*Ts]): ...
 
-In general, some kind of extra constraint is necessary in order
-for the ambiguity to be resolved. This is usually provided by
-an un-concatenated usage of the type variable tuple elsewhere in
-the same signature.
+    def bar(x: Tuple[int, ...]):
+      foo(*x)  # NOT valid
 
-For example, resolving ambiguity in an argument:
+Finally, note that a type variable tuple may *not* be used as the type of
+``**kwargs``. (We do not yet know of a use-case for this feature, so we prefer
+to leave the ground fresh for a potential future PEP.)
 
 ::
 
-    def func(ham: Tuple[*Ts1, *Ts2], spam: Ts2): ...
+    # NOT valid
+    def foo(**kwargs: *Ts): ...
 
-    # Ts1 is bound to Tuple[int], Ts2 to Tuple[str, bool]
-    ham: Tuple[int, str, bool]
-    spam: Tuple[str, bool]
-    func(ham, spam)
+Type Variable Tuples with ``Callable``
+--------------------------------------
 
-In a return type:
+Type variable tuples can also be used in the arguments section of a
+``Callable``:
 
 ::
 
-    def func(ham: Ts1, spam: Ts2) -> Tuple[*Ts1, *Ts2]): ...
-
-    ham: Tuple[int]
-    spam: Tuple[str, bool]
-    # Return type is Tuple[int, str, bool]
-    func(ham, spam)
-
-Note, however, that the same cannot be done with generic classes:
+    class Process:
+      def __init__(
+        target: Callable[[*Ts], Any],
+        args: Tuple[*Ts]
+      ): ...
 
-::
+    def func(arg1: int, arg2: str): ...
+    
+    Process(target=func, args=(0, 'foo'))  # Valid
+    Process(target=func, args=('foo', 0))  # Error
 
-    # No way to add extra constraints about Ts1 and Ts2,
-    # so this is not valid
-    class C(Generic[*Ts1, *Ts2]): ...
+However, note that as of this PEP, if a type variable tuple does appear in
+the arguments section of a ``Callable``, it must appear alone.
+That is, `Type Prefixing` is not supported in the context of ``Callable``.
+(PEP 612 covers the use-cases where this might otherwise be useful.)
```

Comment:
I recommend that you also include a sentence about compatibility with callback prototypes as defined in PEP 544. In particular, TypeVarTuple can be used in the `__call__` signature of a callback prototype.

Reply:
Ooh, I didn't know about callback protocols, nice. Tweaked the last sentence hear to reference them.